File: app/src/types/tunnel.ts
/**
 * 端口转发隧道类型定义
 *
 * 与后端 `crates/ssh-core/src/tunnel.rs` 对齐：
 * - TunnelConfig 字段为 snake_case（Tauri 只转换顶层参数名，不递归转换嵌套字段）
 * - kind 取值为 snake_case 字符串
 *
 * 字段语义按 kind 区分（与 OpenSSH -L/-R/-D 一致）：
 * - local：local_host:local_port 本地监听 → remote_host:remote_port 远端目标
 * - remote：remote_host:remote_port 远端监听（port 0 = 服务器分配，启动后回填）
 *          → local_host:local_port 本地目标
 * - dynamic：local_host:local_port 本地 SOCKS5 监听，remote_* 不使用
 */

/** 隧道类型 */
export type TunnelKind = 'local' | 'remote' | 'dynamic'

/** 隧道配置（ProfileExtra.tunnels 与 tunnel_start 入参共用） */
export interface TunnelConfig {
  /** 隧道 ID（空串由后端生成 UUID；档案内配置在编辑时预生成） */
  id: string
  /** 隧道类型 */
  kind: TunnelKind
  /** 本地地址（local/dynamic=监听地址；remote=本地目标地址） */
  local_host: string
  /** 本地端口 */
  local_port: number
  /** 远端地址（local=远端目标地址；remote=远端监听地址；dynamic 不使用） */
  remote_host: string
  /** 远端端口 */
  remote_port: number
}

/** 隧道运行状态（后端 TunnelStatus::as_str） */
export type TunnelState = 'running' | 'stopped' | 'failed'

/** 隧道列表项（tunnel_list 返回，snake_case） */
export interface TunnelInfo {
  /** 隧道 ID */
  id: string
  /** 所属会话 ID */
  session_id: string
  /** 隧道配置（remote 端口 0 时已回填分配端口） */
  config: TunnelConfig
  /** 状态串 */
  state: TunnelState
  /** 失败原因（仅 failed 时有值） */
  error?: string | null
}

/** `tunnel_event` 事件 payload（camelCase，对应 events.rs TunnelEventPayload） */
export interface TunnelEventPayload {
  /** 隧道 ID */
  tunnelId: string
  /** 会话 ID */
  sessionId: string
  /** 状态串 */
  state: TunnelState
  /** 失败原因（仅 failed 时有值） */
  error?: string | null
}

/** 创建空隧道配置（新建表单/档案编辑的初始行） */
export function emptyTunnelConfig(kind: TunnelKind = 'local'): TunnelConfig {
  return {
    id: crypto.randomUUID(),
    kind,
    local_host: '127.0.0.1',
    local_port: kind === 'dynamic' ? 1080 : 8080,
    remote_host: '127.0.0.1',
    remote_port: 80,
  }
}

File: crates/desktop/src/commands/tunnel.rs
//! # 端口转发命令组
//!
//! 对接前端隧道管理面板：启动 / 停止 / 列出当前会话的 SSH 端口转发隧道
//! （Local / Remote / Dynamic，实现见 ssh-core::tunnel）。
//!
//! 状态变更通过全局事件 `tunnel_event`（[`TunnelEventPayload`]）推送：
//! - 启动成功 / 停止：由本模块在命令路径上 emit
//! - 运行期失败（如 remote 转发因会话断开失效）：由注入 TunnelManager 的
//!   notify 回调在后台 task 中 emit

use std::sync::Arc;

use tauri::{AppHandle, Emitter, State};

use crate::error::AppError;
use crate::events::TunnelEventPayload;
use crate::AppState;
use ssh_core::tunnel::{TunnelConfig, TunnelInfo, TunnelNotify, TunnelStatus};

/// 构造隧道状态变更通知回调：emit `tunnel_event` → 前端刷新隧道列表
fn tunnel_notify<R: tauri::Runtime>(app: &AppHandle<R>) -> TunnelNotify {
    let app = app.clone();
    Arc::new(move |tunnel_id: String, session_id: String, status: TunnelStatus| {
        let _ = app.emit(
            "tunnel_event",
            TunnelEventPayload {
                tunnel_id,
                session_id,
                state: status.as_str().to_string(),
                error: match &status {
                    TunnelStatus::Failed(e) => Some(e.clone()),
                    _ => None,
                },
            },
        );
    })
}

/// 启动隧道，返回 tunnel_id
///
/// `config.id` 为空时由后端生成 UUID（档案自动启动场景由前端预生成，
/// 以便面板把运行中隧道与档案配置对应标注来源）。
/// 监听绑定 / tcpip-forward 请求失败时直接返回错误（隧道未建立）。
#[tauri::command]
pub async fn tunnel_start<R: tauri::Runtime>(
    app: AppHandle<R>,
    state: State<'_, AppState>,
    session_id: String,
    config: TunnelConfig,
) -> Result<String, AppError> {
    let session = state
        .ssh_manager
        .get(&session_id)
        .await
        .ok_or_else(|| AppError::session(format!("会话 {session_id} 不存在或已断开")))?;

    let notify = tunnel_notify(&app);
    let tunnel_id = state
        .tunnels
        .start(session, config, Some(notify))
        .await
        .map_err(|e| AppError::session(format!("{e}")))?;

    let _ = app.emit(
        "tunnel_event",
        TunnelEventPayload {
            tunnel_id: tunnel_id.clone(),
            session_id,
            state: "running".to_string(),
            error: None,
        },
    );
    Ok(tunnel_id)
}

/// 停止隧道（幂等：不存在的 id 视为成功）
#[tauri::command]
pub async fn tunnel_stop<R: tauri::Runtime>(
    app: AppHandle<R>,
    state: State<'_, AppState>,
    tunnel_id: String,
) -> Result<(), AppError> {
    // 先查出 session_id 供事件携带（stop 后句柄已移除）
    let session_id = state
        .tunnels
        .get_info(&tunnel_id)
        .await
        .map(|t| t.session_id);
    state
        .tunnels
        .stop(&tunnel_id)
        .await
        .map_err(|e| AppError::session(format!("{e}")))?;
    if let Some(session_id) = session_id {
        let _ = app.emit(
            "tunnel_event",
            TunnelEventPayload {
                tunnel_id,
                session_id,
                state: "stopped".to_string(),
                error: None,
            },
        );
    }
    Ok(())
}

/// 列出指定会话的隧道（含运行状态，面板刷新用）
#[tauri::command]
pub async fn tunnel_list(
    state: State<'_, AppState>,
    session_id: String,
) -> Result<Vec<TunnelInfo>, AppError> {
    Ok(state.tunnels.list(&session_id).await)
}
